visualizer/semanticsHelper.js

```diff
@@ -6,26 +6,181 @@
   if (typeof exports === 'object') {
     module.exports = initModule;
   } else {
-    initModule(root.ohmEditor, root.CheckedEmitter);
+    initModule(root.ohm, root.ohmEditor, root.CheckedEmitter);
   }
-})(this, function(ohmEditor, CheckedEmitter) {
-  // Exports
-  // -------
-  ohmEditor.semantics = new CheckedEmitter();
-  ohmEditor.semantics.registerEvents({
-    // Emitted after adding an new operation/attribute
-    'add:semanticOperation': ['type', 'name', 'optArguments']
-  });
+})(this, function(ohm, ohmEditor, CheckedEmitter) {
 
-  // Privates
-  // --------
   var semantics = null;
+
+  // `opName` is the semantic operation that current executing, it's null when every
+  // an empty semantics is created.
+  var opName = null;
+
+  // `opArguments` is the operation arguments for the operation with `opName`, could
+  // be empty.
+  var opArguments = null;
+
   ohmEditor.addListener('parse:grammar', function(matchResult, grammar, error) {
     if (grammar && grammar.defaultStartRule) {
       semantics = grammar.semantics();
+      opName = null;
+      opArguments = null;
     }
   });
 
+  ohmEditor.semantics.addListener('change:semanticOperation', function(targetName, optArgs) {
+    if (opName === targetName) {
```

Comment:
I still don't understand why you clear opName if it's equal to targetName...seems a bit odd, and worth a comment explaining why.